opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/measurement_consumer.py

```diff
@@ -90,12 +92,38 @@ def register_asynchronous_instrument(
     def collect(
         self,
         metric_reader: "opentelemetry.sdk.metrics.MetricReader",
+        timeout_millis: float = 10_000,
     ) -> Iterable[Metric]:
+
+        errors = []
+
         with self._lock:
             metric_reader_storage = self._reader_storages[metric_reader]
             # for now, just use the defaults
             callback_options = CallbackOptions()
+            deadline_ns = _time_ns() + timeout_millis * 10**6
             for async_instrument in self._async_instruments:
-                for measurement in async_instrument.callback(callback_options):
-                    metric_reader_storage.consume_measurement(measurement)
+                try:
```

Comment:
If we have a timeout, shouldn't we just immediately bubble up to the caller (since we exceeded the expected timeout)?

Exceptions in the actual async instrument callbacks are already caught and logged here so we don't need to worry about that raising https://github.com/open-telemetry/opentelemetry-python/blob/240de22e8123f39ea18fd9b0fe43c1957fd6dcce/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/instrument.py#L138-L141

Reply:
Fixed